components/eamxx/src/diagnostics/zonal_avg.cpp

```diff
@@ -0,0 +1,189 @@
+#include "diagnostics/zonal_avg.hpp"
+
+#include "share/field/field_utils.hpp"
+
+namespace scream {
+
+void ZonalAvgDiag::compute_zonal_sum(const Field &result, const Field &field, const Field &weight,
+                                     const Field &lat, const ekat::Comm *comm) {
+  auto result_layout       = result.get_header().get_identifier().get_layout();
+  const int num_zonal_bins = result_layout.dim(0);
+  const int ncols          = field.get_header().get_identifier().get_layout().dim(0);
+  const Real lat_delta     = sp(180.0) / num_zonal_bins;
+
+  auto weight_view = weight.get_view<const Real *>();
+  auto lat_view    = lat.get_view<const Real *>();
+  using KT         = ekat::KokkosTypes<DefaultDevice>;
+  using TeamPolicy = Kokkos::TeamPolicy<Field::device_t::execution_space>;
+  using TeamMember = typename TeamPolicy::member_type;
+  using ESU        = ekat::ExeSpaceUtils<typename KT::ExeSpace>;
+  switch (result_layout.rank()) {
+  case 1: {
+    auto field_view        = field.get_view<const Real *>();
+    auto result_view       = result.get_view<Real *>();
+    TeamPolicy team_policy = ESU::get_default_team_policy(num_zonal_bins, ncols);
+    Kokkos::parallel_for(
+        "compute_zonal_sum_" + field.name(), team_policy, KOKKOS_LAMBDA(const TeamMember &tm) {
+          const int lat_i      = tm.league_rank();
+          const Real lat_lower = sp(-90.0) + lat_i * lat_delta;
+          const Real lat_upper = lat_lower + lat_delta;
+          Kokkos::parallel_reduce(
+              Kokkos::TeamVectorRange(tm, ncols),
+              KOKKOS_LAMBDA(int i, Real &val) {
```

Comment:
This (and similarly below) is the issue on CUDA: you should only use KOKKOS_LAMBDA for the _outermost_ functor. The macro add some decorator that don't make sense when you are already in a || region.

Reply:
@mahf708 , I'll take a look a the quickstart doc, thanks for point me to it!

@bartgol , I see the macro definition now... and why the by-copy capture is not the way to go: thanks for the heads up!

Reply:
I applied the change to the parallel reduction functors in [2d55ab7](https://github.com/E3SM-Project/E3SM/pull/7261/commits/2d55ab766d1ebc22b0c665256afac9728d6de162).  Before and after the change, I'm unable to reproduce any compilation errors (using the `test-scream-eamxx` script or a direct "cmake + make" approach).

Reply:
It did just occur to me that I'm testing on CPU-only machine... which would explain why I'm not seeing a CUDA related issue.  I'll look at testing on a machine with GPUs (like perlmutter) but in the meantime, hopefully the changes Luca suggested address the compilation problems.

Reply:
>Okay if I push to your branch?
That works for me: thanks for taking a look.